README.md
## socioFetcher

A tool built for simplifying socio-economic data downloading process, integrating python API wrappers of [BLS](https://www.bls.gov/), [BEA](https://www.bea.gov/), and [Census](https://www.census.gov/). The tool also allows users to mapping the data in an interactive map with [ipyleaflet](https://github.com/jupyter-widgets/ipyleaflet) in Jupyter notebook. 

### Demo

#### Download and export Raw data
![raw data download](RawData.gif)

#### Visualize it interactively in Jupyter notebook
![Visualization](map.gif)

## Installation
```
git clone https://github.com/tianxie1995/socio-fetcher.git
cd socio-fetcher
pip install . 
```

### Test
```
pytest -v
 ```

## Use example
Complete examples are located in [example folder](example)
### Download raw data
```
# Customize Config 
myConfig = Config()
myConfig.BLS.API_KEY = "YOUR_BLS_KEY"
myConfig.BLS.TABLE_NUMBER = ["ENU"]
myConfig.BLS.DATA_TYPE = ["1"]
myConfig.BLS.SIZE = ["0"]
myConfig.BLS.OWNERSHIP = ["5"]
myConfig.BLS.NAICS_CODE_LIST = {
            "10":	"10 Total, all industries",
            "101":	"101 Goods-producing",
            "102"	: "102 Service-providing"
        }

myConfig.BEA.API_KDY = "YOUR_BEA_KEY"
myConfig.BEA.LINE_CODE = ["3"]
myConfig.BEA.TABLE_NAME = ["CAINC1"]
myConfig.BEA.YEAR = ["ALL"]

datasets = ["BLS", "BEA"]
fips = {"26093": "Livingston,MI",
        "26099": "Macomb,MI",
        "26115": "Monroe,MI",
        "26125": "Oakland,MI",
        "26147": "St. Clair,MI",
        "26161": "Washtenaw,MI",
        "26163": "Wayne,MI"}
dl = Downloader(datasets, list(fips.keys()), config=myConfig)
dl.download()
dl.export("data/")
```
### Download summary by geography & dataset
```
# Customize Config 
myConfig = Config()
...
...

datasets = ["BLS", "BEA"]
fips = {"26093": "Livingston,MI",
        "26099": "Macomb,MI",
        "26115": "Monroe,MI",
        "26125": "Oakland,MI",
        "26147": "St. Clair,MI",
        "26161": "Washtenaw,MI",
        "26163": "Wayne,MI"}
dl = Downloader(datasets, list(fips.keys()), config=myConfig)
dl.download()
# export summary by geography
dl.export("summary/", summarize=True, by="geography")
# export summary by dataset
dl.export("summary/", summarize=True, by="dataset")
```

### Visualization with Jupyter notebook
```
# Customize Config 
myConfig = Config()
...
...

datasets = ["BLS", "BEA"]
fips = {"26093": "Livingston,MI",
        "26099": "Macomb,MI",
        "26115": "Monroe,MI",
        "26125": "Oakland,MI",
        "26147": "St. Clair,MI",
        "26161": "Washtenaw,MI",
        "26163": "Wayne,MI"}
dl = Downloader(datasets, list(fips.keys()), config=myConfig)
dl.download()
mapView = dl.mapping("BLS")
mapView.show(
        center=(42.346814, -83.319304), # specify map center
        zoom=8, # specify map zoom level 
        **kwargs # other map attributes available at https://ipyleaflet.readthedocs.io/en/latest/api_reference/map.html
)
```

## Report bugs and contribute to this tool
Please use [Github Issues]() to report bugs and contribute to this tool. 


## License 
This tool is licensed under [BSD License](LICENSE), see LICENSE for more detail. 


CHANGELOG.rst
Change Log
==============

Expected use
--------------
::

    # config
    from socioFetcher import Config
    myConfig = Config()
    myConfig.show() # show current config
    myConfig.Global.areaCode = {...}
    myConfig.BLS.key = "key"
    myConfig.Census.key = "key"
    myConfig.BEA.key = "key"
    ...
    # download
    from socioFetcher import Downloader
    downloader = Downloader(["BEA", "BLS"], ["26193"], config=myConfig)
    downloader.download()
    # summarize
    downloader.sumerize(by="geography")
    # mapping in Jupyter notebook
    m = downloader.mapping(dataset="BEA")
    m.show()
    # export
    downloader.export("path/to/save/data",**kwargs)

Todo
--------------
- Rewrite pep downloader: https://api.census.gov/data/2018/pep/components?key=a1b79f5105b689bd9c4ed357de83130393b6dec7&get=BIRTHS,PERIOD_DESC&for=county:125&in=state:26&PERIOD_CODE=9
- Update testing for geodataframe 

Log
--------------
- 01/03/20 update GDP download
- 10/29 Fix BEA column name bug where same column name exist
- 09/04 Reformat ACS Config to include options to choose dataset
- 09/03 Reformat ACS Config and refactor ACS downloader
- 08/05 
    - Add BLS unEmployment data download feature,
    - add related config setting
    - Add export to json options
- 08/01 
    - Add gettingstarted and installation pages in docs
    - Remove all credentials in the config and edit .gitigonore to exclude test folder
    - Transfer the repo to SEMCOG account
- 07/31 Add doc session in root dir, add docstring in modules
- 07/30 
    - Improving BLS downloading and bug fixed
    - Added and tested MapView module, mapping data to the map
- 07/29 Finished and test mapping module. 91% cov
    - 📝Add year, attr select and data display widget in the map
    - 📝Add progress bar when downloading data
    - 🖇Edit test to reflect changes
- 07/26 📝Add func to produce choro data by attr and year
- 07/25 📝Add download geojson from Census REST feature, choro map, iteractive map, handling click.
- 07/24 
    - 🖇Rename Config api variables for better user understanding 
    - 🖇Remove redundant Config variable 
    - 🐞Fix a bug which cause produce repetitive record in DataFrame when fetching ACS data
    - 
- 07/23 Refactor downloader with request.Session
    - 🖇Refactor downloader using request.Session, improve profermance
- 07/22 🖇Refactor config.py to include a class to hold all config attributes
        🖇Edit testing files to reflect the changes in config
- 07/19 📝 Add unit test for downloader, add func test for download,
        fix bugs in BEA download
- 07/18 📝 Add summary and export feature to Downloader
    - Intergrate summary.py into Downloader
        - Sumerize by either geography or dataset
    - 📝Add export function to Downloader
        - export either individual file or summary
            file by either geography or dataset
    - 📝Add unit testing of geodataframe

- 07/17 🧹Rewrite download.py in Class format downloader.py
    - Downloader accept options in constructor
    - Use Downloader.download() func download data and save to Downloader.data
- 07/16 Set up the pakeage required files


example/downloadRawData.py
from socioFetcher import Config, Downloader

# Customize Config
myConfig = Config()
myConfig.BLS.API_KEY = "YOUR_BLS_KEY"
myConfig.BLS.TABLE_NUMBER = ["ENU"]
myConfig.BLS.DATA_TYPE = ["1"]
myConfig.BLS.SIZE = ["0"]
myConfig.BLS.OWNERSHIP = ["5"]
myConfig.BLS.NAICS_CODE_LIST = {
    "10":	"10 Total, all industries",
    "101":	"101 Goods-producing",
    "102"	: "102 Service-providing"
}

myConfig.BEA.API_KDY = "YOUR_BEA_KEY"
myConfig.BEA.LINE_CODE = ["3"]
myConfig.BEA.TABLE_NAME = ["CAINC1"]
myConfig.BEA.YEAR = ["ALL"]

datasets = ["BLS", "BEA"]
fips = {"26093": "Livingston,MI",
        "26099": "Macomb,MI",
        "26115": "Monroe,MI",
        "26125": "Oakland,MI",
        "26147": "St. Clair,MI",
        "26161": "Washtenaw,MI",
        "26163": "Wayne,MI"}
dl = Downloader(datasets, list(fips.keys()), config=myConfig)
dl.download()
dl.export("data/")
